File: src/main/java/com/github/nsc/de/shake/util/CompilerError.java
package com.github.nsc.de.shake.util;

import com.github.nsc.de.shake.lexer.characterinput.position.Position;
import com.github.nsc.de.shake.lexer.characterinput.position.PositionMap;

public class CompilerError extends Error {

    private final String name;
    private final String details;
    private final Position start;
    private final Position end;
    private final ErrorMarker marker;
    public static int maxLength = 60;
    private static Position start_zw;
    private static Position end_zw;

    private CompilerError(String message, ErrorMarker marker, String name, String details, Position start, Position end) {
        super(String.format("%s: %s%n%n%s%n", message, details, marker));
        this.name = name;
        this.details = details;
        this.start = start;
        this.end = end;
        this.marker = marker;
    }

    private CompilerError(String message, ErrorMarker marker, String name, String details, PositionMap map, int start, int end) {
        super(String.format("%s: %s%n%n%s%n", message, details, marker));
        this.name = name;
        this.details = details;
        this.start = map.resolve(start);
        this.end = map.resolve(end);
        this.marker = marker;
    }

    public CompilerError(String message, String name, String details, Position start, Position end) {
        this(message, createPositionMarker(CompilerError.maxLength, start, end), name, details, start, end);
    }

    public CompilerError(String message, String name, String details, PositionMap map, int start, int end) {
        this(message, createPositionMarker(CompilerError.maxLength, start_zw = map.resolve(start), end_zw = map.resolve(end)), name, details, start_zw, end_zw);
        start_zw = end_zw = null;
    }

    public String getName() {
        return name;
    }
    public String getDetails() {
        return details;
    }
    public Position getStart() {
        return start;
    }
    public Position getEnd() {
        return end;
    }
    public ErrorMarker getMarker() { return marker; }

    @Override
    public String toString() {
        return getMessage();
    }

    @Override
    public void printStackTrace() {
        super.printStackTrace();
    }

    private static ErrorMarker createPositionMarker(int maxLength, Position pos1, Position pos2) {

        // Check requirements
        if(pos1.getSource() != pos2.getSource()) throw new Error("The two have to be located in the same source");
        if(pos1.getLine() != pos2.getLine()) throw new Error("The two positions that should be marked have to be in the same line");
        if(pos1.getColumn() > pos2.getColumn()) throw new Error("Position 1 must be before Position 2");



        // Line start (linenumber + 2 spaces)
        String line_str = pos1.getLine() + "  ";

        // Length of the highlighted section
        int highlighted = pos2.getColumn() - pos1.getColumn() + 1;

        // The maximum amount of characters that will be shown around the highlighted section
        int maxAround = maxLength - highlighted - line_str.length();
        int before = maxAround / 2 + maxAround % 2;
        int after = maxAround / 2;




        // The available tokens before the highlighted section
        int before2 = pos1.getColumn() - 1;

        // The available tokens after the highlighted section
        int after2 = pos2.getSource().getAfterInLine(pos2);

        // Take the smallest value and use it
        int real_before = smallest(before, before2);
        int real_after = smallest(after, after2 + 1);

        // Get the differences (to display if there are tokens that can't be displayed)
        int before_dif = before2 - real_before;
        int after_dif = after2 - real_after;

        // Resolve numbers if there is a non-displayed part
        if(before_dif > 0) {
            int baseLen = String.valueOf(before_dif).length();
            int len = baseLen + 4;
            if(String.valueOf(len).length() != baseLen) real_before -= ++len;
            else real_before -= len;
            before_dif += len;
        }
        if(after_dif > 0) {
            int baseLen = String.valueOf(after_dif).length();
            int len = baseLen + 4;
            if(String.valueOf(len).length() != baseLen) real_after -= ++len;
            else real_after -= len;
            after_dif += len;
        }


        // The start of the line
        String start = line_str
                + (before_dif > 0 ? ("+" + before_dif + "...") : "")
                + String.valueOf(pos1.getSource().getSource().get(pos1.getIndex() - real_before, pos1.getIndex()));

        // The end of the line
        String end = String.valueOf(pos1.getSource().getSource().get(pos2.getIndex()+1, pos2.getIndex() + real_after))
                + (after_dif > 0 ? ("...+" + after_dif) : "");

        // Generate end-string
        return new ErrorMarker(
                pos1.getSource().getLocation() + ':' + pos1.getLine() + ':' + pos1.getColumn(),
                start + Formatting.INVERT + Formatting.FGColor.RED +
                        String.valueOf(pos1.getSource().getSource().get(pos1.getIndex(), pos2.getIndex()+1)) + Formatting.RESET + end,
                start + String.valueOf(pos1.getSource().getSource().get(pos1.getIndex(), pos2.getIndex()+1)) + end,
                strRepeat(' ', start.length())  + strRepeat('^', highlighted));

    }

    private static String strRepeat(char c, int number) {
        StringBuilder string = new StringBuilder();
        for(int i = 0; i < number; i++) string.append(c);
        return string.toString();
    }

    private static int smallest(int ...arr) {
        int smallest = arr[0];
        for(int i = 1; i < arr.length; i++) if(arr[i] < smallest) smallest = arr[i];
        return smallest;
    }

    public static class ErrorMarker {

        private final String source;
        private final String colorPreview;
        private final String preview;
        private final String marker;

        public ErrorMarker(String source, String colorPreview, String preview, String marker) {
            this.source = source;
            this.colorPreview = colorPreview;
            this.preview = preview;
            this.marker = marker;
        }

        public String getSource() { return source; }
        public String getPreview() { return preview; }
        public String getColorPreview() { return colorPreview; }
        public String getMarker() { return marker; }

        @Override
        public String toString() {
            return "at " + getSource() + System.lineSeparator() + getColorPreview() + System.lineSeparator() + getMarker();
        }
    }
}

File: util/src/main/kotlin/com/github/nsc/de/shake/util/CompilerError.kt
package com.github.nsc.de.shake.util

import com.github.nsc.de.shake.lexer.characterinput.position.Position
import com.github.nsc.de.shake.lexer.characterinput.position.PositionMap
import java.lang.StringBuilder
import com.github.nsc.de.shake.util.Formatting.FGColor
import java.lang.Error

@Suppress("unused")
open class CompilerError : Error {
    val name: String
    val details: String
    val start: Position
    val end: Position
    val marker: ErrorMarker

    private constructor(
        message: String, marker: ErrorMarker, name: String, details: String, start: Position, end: Position,
        cause: Throwable
    ) : super(String.format("%s: %s%n%n%s%n", message, details, marker), cause) {
        this.name = name
        this.details = details
        this.start = start
        this.end = end
        this.marker = marker
    }

    private constructor(
        message: String, marker: ErrorMarker, name: String, details: String, map: PositionMap, start: Int, end: Int,
        cause: Throwable
    ) : super(String.format("%s: %s%n%n%s%n", message, details, marker), cause) {
        this.name = name
        this.details = details
        this.start = map.resolve(start)
        this.end = map.resolve(end)
        this.marker = marker
    }

    constructor(
        message: String, name: String, details: String, start: Position, end: Position,
        cause: Throwable
    ) : this(message, createPositionMarker(maxLength, start, end), name, details, start, end, cause)

    constructor(
        message: String, name: String, details: String, map: PositionMap, start: Int, end: Int,
        cause: Throwable
    ) : this(message, createPositionMarker(maxLength, map.resolve(start).also { start_zw = it },
        map.resolve(end).also {
            end_zw = it
        }), name, details, start_zw!!, end_zw!!, cause
    ) {
        end_zw = null
        start_zw = end_zw
    }

    private constructor(
        message: String,
        marker: ErrorMarker,
        name: String,
        details: String,
        start: Position,
        end: Position
    ) : super(
        String.format("%s: %s%n%n%s%n", message, details, marker)
    ) {
        this.name = name
        this.details = details
        this.start = start
        this.end = end
        this.marker = marker
    }

    private constructor(
        message: String,
        marker: ErrorMarker,
        name: String,
        details: String,
        map: PositionMap,
        start: Int,
        end: Int
    ) : super(
        String.format("%s: %s%n%n%s%n", message, details, marker)
    ) {
        this.name = name
        this.details = details
        this.start = map.resolve(start)
        this.end = map.resolve(end)
        this.marker = marker
    }

    constructor(message: String, name: String, details: String, start: Position, end: Position) : this(
        message, createPositionMarker(
            maxLength, start, end
        ), name, details, start, end
    ) {
    }

    constructor(message: String, name: String, details: String, map: PositionMap, start: Int, end: Int) : this(
        message, createPositionMarker(maxLength, map.resolve(start).also { start_zw = it },
            map.resolve(end).also {
                end_zw = it
            }), name, details, start_zw!!, end_zw!!
    ) {
        end_zw = null
        start_zw = end_zw
    }

    override fun toString(): String {
        return message!!
    }

    override fun printStackTrace() {
        super.printStackTrace()
    }

    class ErrorMarker(val source: String, val colorPreview: String, val preview: String, val marker: String) {
        override fun toString(): String {
            return "at " + source + System.lineSeparator() + colorPreview + System.lineSeparator() + marker
        }
    }

    companion object {
        var maxLength = 60
        private var start_zw: Position? = null
        private var end_zw: Position? = null
        private fun createPositionMarker(maxLength: Int, pos1: Position, pos2: Position): ErrorMarker {

            // Check requirements
            if (pos1.source != pos2.source) throw Error("The two have to be located in the same source")
            if (pos1.line != pos2.line) throw Error("The two positions that should be marked have to be in the same line")
            if (pos1.column > pos2.column) throw Error("Position 1 must be before Position 2")


            // Line start (linenumber + 2 spaces)
            val line_str = pos1.line.toString() + "  "

            // Length of the highlighted section
            val highlighted = pos2.column - pos1.column + 1

            // The maximum amount of characters that will be shown around the highlighted section
            val maxAround = maxLength - highlighted - line_str.length
            val before = maxAround / 2 + maxAround % 2
            val after = maxAround / 2


            // The available tokens before the highlighted section
            val before2 = pos1.column - 1

            // The available tokens after the highlighted section
            val after2 = pos2.source.getAfterInLine(pos2)

            // Take the smallest value and use it
            var real_before = smallest(before, before2)
            var real_after = smallest(after, after2 + 1)

            // Get the differences (to display if there are tokens that can't be displayed)
            var before_dif = before2 - real_before
            var after_dif = after2 - real_after

            // Resolve numbers if there is a non-displayed part
            if (before_dif > 0) {
                val baseLen = before_dif.toString().length
                var len = baseLen + 4
                real_before -= if (len.toString().length != baseLen) ++len else len
                before_dif += len
            }
            if (after_dif > 0) {
                val baseLen = after_dif.toString().length
                var len = baseLen + 4
                real_after -= if (len.toString().length != baseLen) ++len else len
                after_dif += len
            }


            // The start of the line
            val start = (line_str
                    + (if (before_dif > 0) "+$before_dif..." else "")
                    + String(pos1.source.source[pos1.index - real_before, pos1.index]).replace("\t".toRegex(), " "))

            // The end of the line
            val end = (String(pos1.source.source[pos2.index + 1, pos2.index + real_after]).replace("\t".toRegex(), " ")
                .replace("\n".toRegex(), " ")
                    + if (after_dif > 0) "...+$after_dif" else "")

            // Generate end-string
            return ErrorMarker(
                pos1.source.location + ':' + pos1.line + ':' + pos1.column,
                start + Formatting.INVERT + FGColor.RED + String(pos1.source.source[pos1.index, pos2.index + 1]) + Formatting.RESET + end,
                start + String(pos1.source.source[pos1.index, pos2.index + 1]) + end,
                strRepeat(' ', start.length) + strRepeat('^', highlighted)
            )
        }

        private fun strRepeat(c: Char, number: Int): String {
            val string = StringBuilder()
            for (i in 0 until number) string.append(c)
            return string.toString()
        }

        private fun smallest(vararg arr: Int): Int {
            var smallest = arr[0]
            for (i in 1 until arr.size) if (arr[i] < smallest) smallest = arr[i]
            return smallest
        }
    }
}